lib/sycamore/sycamore/writers/base.py

```diff
@@ -0,0 +1,76 @@
+from dataclasses import dataclass
+
+from abc import ABC, abstractmethod
+from typing import Callable
+
+from sycamore.data.document import Document
+from sycamore.plan_nodes import Node, Write
+from sycamore.transforms.map import MapBatch
+from sycamore.utils.ray_utils import check_serializable
+
+
+class BaseDBWriter(MapBatch, Write):
+
+    # Type param for the client
+    class Client(ABC):
+        @classmethod
+        @abstractmethod
+        def from_client_params(cls, params: "BaseDBWriter.ClientParams") -> "BaseDBWriter.Client":
+            pass
+
+        @abstractmethod
+        def write_many_records(self, records: list["BaseDBWriter.Record"], target_params: "BaseDBWriter.TargetParams"):
+            pass
+
+        @abstractmethod
+        def create_target_idempotent(self, target_params: "BaseDBWriter.TargetParams"):
+            pass
+
+        @abstractmethod
+        def get_existing_target_params(self, target_params: "BaseDBWriter.TargetParams") -> "BaseDBWriter.TargetParams":
+            pass
+
+    # Type param for the objects to write to the db
+    class Record(ABC):
+        @classmethod
+        @abstractmethod
+        def from_doc(cls, document: Document, target_params: "BaseDBWriter.TargetParams") -> "BaseDBWriter.Record":
+            pass
+
+    # Type param for the object used to configure a new index if necessary
```

Comment:
a new target

if you think it's needed: , e.g. opensearch or pinecone index, weviate collection, filesystem directory, or s3 bucket.